[알린]Store step7 - PR

jinios opened the pull request:
상세화면 구현과 주문하기 기능 구현했습니다!!

jinios committed:
Docs: Update README

jinios committed:
Rename CategoryProtocol to URLDataProtocol

jinios committed:
Feat: Add and basic implement DetailHash class #13

jinios committed:
Feat: Make model for DetailHash #13

Class explanation
- DetailHash: model object of each itemCell for indicate hashURL and
product detail info
- HashDataSetter: Resume DataTask for parsing data getting from detail
hash URL
- ItemHashData, DetailItemInfo: Codable object by parsing detail hash

jinios committed:
Test: Detail hash data test

jinios committed:
Feat: Add itemData() #13

jinios committed:
Feat: Add ItemViewController #14
- For implement detail item view

jinios committed:
Feat: Add itemTitle property #13

jinios committed:
Feat: Implement ItemViewController #14
- Set properties by DetaillHash model
- Adjust layout and modify constraint

jinios committed:
Feat: Implement push itemDetail view when cell tapped #15

jinios committed:
Feat: Implement download thumbnail images #16
- Add thumbnailImages variable
- Implement download thumbnail images

jinios committed:
Feat: Implement Thumbnail scrollview #16

- Add Notification: thumbnailDownloaded - post when thumbnail image
downloading complete
- Add selector: setThumbnailScrollView() - calculate frames and set
scrollview
- Rename variables - thumbnailScrollView

jinios committed:
Feat: Implement detailScrollView #17

- Set detail section info on vertical scrollView

jinios committed:
Feat: Add IBAction method for order button #18
- First Basic implement for order feature in ItemViewController
- Add OrderDelegate protocol
- Connect IBAction method with order button

jinios committed:
Feat: Modify AutoLayout

jinios committed:
Feat: Make OrderItem struct - item ordered #18

- OrderItem struct represents the item ordered
- contains the title and price of the item user purchased

jinios committed:
Feat: Implement order()- slack incoming webhook #18

- Implement URLRequest and post request for using incoming webhook
- User can order an item by tapping order button on ItemViewController's
view

jinios committed:
Refactor: Change string constant to enum #18

jinios committed:
Refactor: Modify duplicated codes #16,#17

jinios committed:
Refactor: Rename enum case- httpPostRequest #18

jinios committed:
Docs- Update README - step7

godrm commented on StoreApp/StoreApp/ItemViewController.swift:
```diff
@@ -0,0 +1,155 @@
+                    }
+                }
+            })
+        }
```
이 메소드는 이전 단계에서 했던것처럼 네트워크에 접근하는 모델로 분리하는게 어떨까요? 
심지어 노티를 보내고 있군요!

jinios replied:
이것도 ImageSetter안에서 다 class 함수로 처리하고싶었는데...
그러면 downloadItemImages 함수에서 UIImageView 타입을 사용하기때문에 모델객체인 ImageSetter파일에 UIKit을 import해야할거같은데 그게 역할분담이 깨지는거같아서 뷰컨트롤러에서 구현했습니다... :( 
조금 더 고민해보겠습니다 !

godrm replied:
그렇다면 모델 역할과 뷰 역할이 뒤섞여 있는게 아닐까요?

jinios replied:
네 그래서 ImageSetter에서 다 처리한다면 UIImageView를 만드는 데이터만 뷰컨트롤러에게 전달하고 실제로 UIImageView를 만드는 객체는 뷰컨트롤러가 되도록 분리해야할것같습니다!...?

godrm commented on StoreApp/StoreApp/ItemViewController.swift:
```diff
@@ -0,0 +1,155 @@
+                                           width: self.detailScrollView.frame.width, height: self.detailScrollView.frame.height)
+            detailScrollView.addSubview(images[i])
+        }
+    }
```
setThumbnailScrollView() 나 setDetailScrollView() 에서 하는 작업은 결국 내부 뷰에 서브뷰를 추가하는 작업인데 
이 부분을 ViewController에서 하지 않는다면 어떻게 분리할 수 있을까요?

godrm commented on StoreApp/StoreApp/Model/OrderDelegate.swift:
```diff
@@ -0,0 +1,43 @@
+        URLSession.shared.dataTask(with: request).resume()
+    }
+    
+}
```
별도 다른 객체로 분리하지 않고 extension 으로 확장한 이유는 무엇인가요? 
어떤 장단점이 있길래 선택했는지 궁금합니다.

jinios replied:
지금 프로젝트에서는 order함수의 동작이 정해져있고 커스텀이 필요하지 않아서(url을 바꾸거나 http요청을 바꾸거나 하는 변경사항이 없음), extension에 기본 구현만 해놓으면 델리게이트를 채택한 객체는 그대로 함수를 가져다 쓸 수 있기때문에 사용했습니다.
만약 델리게이트 객체를 만들어서 해당 객체에서 구현하는것으로 분리한다면 커스텀하기가 좋아져서, 델리게이트 객체마다 url이나 httpMethod를 다르게 부여해서 필요에 따라 다르게 동작하게 구현하기에 편할 것 같습니다.

godrm commented on StoreApp/StoreAppUnitTests/StoreAppUnitTests.swift:
```diff
@@ -9,6 +9,19 @@
+        }
+
+    }
+
```
비동기 작업에 대한 테스트 코드를 좋네요!

jinios committed:
Refactor: Make custom scrollview class #19

- ItemDetailScrollView: custom class to implement scrollview in
ItemViewController

jinios committed:
Refactor:Move download method to ImageSetter #20

- Move downloadDetailImages() from ItemViewController to ImageSetter
- Send notification from model(ImageSetter) to
controller(ImageViewController)
- Rename method: downloadItemImages() to downloadDetailImages()

jinios commented:
피드백 주신부분 수정했습니다.
- 뷰컨트롤러에서 사용하는 스크롤뷰의 타입을 커스텀 클래스로 변경하였습니다.
- ItemViewController에서 뷰를 다운로드하던 동작을 ImageSetter로 옮겨 역할을 분리했습니다.

godrm commented:
Custom ScrollView를 만들때 상속을 받는 방법도 있지만, UIView 인데 ScrollView를 포함하도록 만드는 방법도 있습니다.
두 가지 방식을 언제 어떤 기준으로 선택하는게 좋을까요?
카드게임앱을 만들때도 비슷한 고민을 하지 않았던가요?

godrm commented:
일단 머지할테니 위에 피드백에 대한 답변은 남겨주세요.
